### PR TITLE
Allow Datapack Registry to be configured by RegistryBuilder 

### LINKED ISSUE #1158: [Feature] Datapack Registry Callbacks
Unless I am mistaken, there currently is no way to modify datapack registries in the API. I'm wondering if its planned or if its in the works?

### PR DESCRIPTION
**Pull request by Argon4W**
Modify the RegistryDataLoader$RegistryData and RegistryBuilder to allow datapack registries to be configured by RegistryBuilder.

### CODE REVIEW

**Comment by Matyrobbrt** [comment]
Why do you need to configure dp regs with registrybuilder?
I'm not sure this is the best approach, instead `ModifyRegistriesEvent` should also fire for data pack registries.

**Comment by Argon4W** [comment]
> Why do you need to configure dp regs with registrybuilder? I'm not sure this is the best approach, instead `ModifyRegistriesEvent` should also fire for data pack registries.

Sorry but I noticed that the doc of the ModifyRegistriesEvent said that it cannot be used to modify datapack registries. 

Also, even if this event works, it cannot add default value for the datapack registry cuz the generated registry is MappedRegistry instead of DefaultedMappedRegistry.

**Comment by Matyrobbrt** [comment]
Ah, so your usecase is adding default values to the registry? I suppose that's fine, but if you add the consumer to the registry data, please make it the last parameter of the record.

**Comment by Argon4W** [comment]
> Ah, so your usecase is adding default values to the registry? I suppose that's fine, but if you add the consumer to the registry data, please make it the last parameter of the record.

Okay, I just added a new commit to make the consumer the last parameter.

File: src/main/java/net/neoforged/neoforge/registries/RegistryBuilder.java [hunk 1]
```diff
@@ -102,15 +102,19 @@ public RegistryBuilder<T> disableRegistrationCheck() {
      * @see NewRegistryEvent#register(Registry)
      */
     public Registry<T> create() {
+        return create(Lifecycle.stable(), false);
+    }
+
+    public Registry<T> create(Lifecycle lifecycle, boolean datapackRegistry) {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/registries/RegistryBuilder.java]: There is no need to add this new parameter, just call `disableRegistrationCheck()`.

File: src/main/java/net/neoforged/neoforge/registries/DataPackRegistryEvent.java [hunk 2]
```diff
@@ -74,6 +75,29 @@ public <T> void dataPackRegistry(ResourceKey<Registry<T>> registryKey, Codec<T>
             this.registryDataList.add(new DataPackRegistryData<>(new RegistryDataLoader.RegistryData<>(registryKey, codec, false), networkCodec));
         }
 
+        /**
+         * Registers the registry key as a datapack registry with a given {@link RegistryBuilder} to configure, which will cause data to be loaded from
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/registries/DataPackRegistryEvent.java]:
```suggestion
         * Registers the registry key as a datapack registry with a {@link RegistryBuilder} configurator, which will cause data to be loaded from
```

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/registries/RegistryBuilder.java (hunk 1 above)
Comment (Argon4W) [reply src/main/java/net/neoforged/neoforge/registries/RegistryBuilder.java]: I have added a new commit to resolve this

File: src/main/java/net/neoforged/neoforge/registries/DataPackRegistryEvent.java (hunk 2 above)
Comment (Argon4W) [reply src/main/java/net/neoforged/neoforge/registries/DataPackRegistryEvent.java]: I have added a new commit to resolve this

File: patches/net/minecraft/resources/RegistryDataLoader.java.patch [hunk 3]
```diff
@@ -46,3 +46,26 @@
  
          for (Entry<ResourceLocation, Resource> entry : filetoidconverter.listMatchingResources(p_321535_).entrySet()) {
              ResourceLocation resourcelocation = entry.getKey();
+@@ -304,13 +_,20 @@
+         void apply(RegistryDataLoader.Loader<?> p_321864_, RegistryOps.RegistryInfoLookup p_321656_);
+     }
+ 
+-    public static record RegistryData<T>(ResourceKey<? extends Registry<T>> key, Codec<T> elementCodec, boolean requiredNonEmpty) {
++    public static record RegistryData<T>(ResourceKey<? extends Registry<T>> key, Codec<T> elementCodec, boolean requiredNonEmpty, java.util.function.Consumer<net.neoforged.neoforge.registries.RegistryBuilder<T>> consumer) {
```
Comment (Matyrobbrt) [inline patches/net/minecraft/resources/RegistryDataLoader.java.patch]: Please rename the new parameter to `registryBuilderConsumer`.

**Review by Matyrobbrt — APPROVED** [review]

**Comment by Shadows-of-Fire** [comment]
I don't think letting a datapack registry be a `DefaultedMappedRegistry` is a good idea, since by virtue of being a data registry, data objects may be removed (including the default object) which will invalidate the nonnull contract of `DefaultedMappedRegistry#get(ResourceLocation)`.

Allowing the configuration through registry builder satisfies requirements from #1158, which is (or rather, might be?) useful, but use of a default key is a footgun.

**Comment by Argon4W** [comment]
> I don't think letting a datapack registry be a `DefaultedMappedRegistry` is a good idea, since by virtue of being a data registry, data objects may be removed (including the default object) which will invalidate the nonnull contract of `DefaultedMappedRegistry#get(ResourceLocation)`.
> 
> Allowing the configuration through registry builder satisfies requirements from #1158, which is (or rather, might be?) useful, but use of a default key is a footgun.

Indeed, the default object may be removed. But it also allow datapacks to change a hardcoded default state, making the datapack behavior more flexible. Also, modifying default object isn’t quite common, and the default value is disabled at default, so if you want to add a default value, you should declare it explicitly and take the risk.

**Comment by Argon4W** [comment]
> I don't think letting a datapack registry be a `DefaultedMappedRegistry` is a good idea, since by virtue of being a data registry, data objects may be removed (including the default object) which will invalidate the nonnull contract of `DefaultedMappedRegistry#get(ResourceLocation)`.
> 
> Allowing the configuration through registry builder satisfies requirements from #1158, which is (or rather, might be?) useful, but use of a default key is a footgun.

BTW, adding default value can prevent the usage of unsafe null value or Optional as default value. Though I can use Holder.direct() and RegistryFileCodec/ByteBufCodecs.holder to achieve similar behavior like the default value, it just feel hacky and the default value has no key to represent

**Comment by Shadows-of-Fire** [comment]
Eh, well, people can use it at their own risk. It isn't really any different than just calling `.holderOrThrow` on a datapack registry, which people are almost certainly doing.

**Pull request merged**